src/codec.rs

```diff
@@ -1099,13 +1100,36 @@ impl Decode for bool {
 	}
 }
 
+impl Encode for Duration {
+	fn size_hint(&self) -> usize {
+		mem::size_of::<u64>() + mem::size_of::<u32>()
+	}
+
+	fn encode(&self) -> Vec<u8> {
+		let secs = self.as_secs();
+		let nanos = self.subsec_nanos();
+		(secs, nanos).encode()
+	}
+}
+
+impl Decode for Duration {
+	fn decode<I: Input>(input: &mut I) -> Result<Self, Error> {
+		let (secs, nanos) = <(u64, u32)>::decode(input)?;
+		if nanos > A_BILLION {
```

Comment:
if nanos is a billion it must be one second then
```suggestion
		if nanos >= A_BILLION {
```